modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C

```diff
@@ -0,0 +1,194 @@
+  bool converged = performExplicitSolve(mass_matrix);
+
+  // Update the solution
+  *_nonlinear_implicit_system->solution = _nl.solutionOld();
+  *_nonlinear_implicit_system->solution += _solution_update;
+
+  _nonlinear_implicit_system->nonlinear_solver->converged = converged;
+
+  // Applying preset BCs
+  setPresetBCs();
+}
+
+void
+CentralDifferenceDirect::setPresetBCs()
+{
+  // Apply preset nodal BC's
+  NumericVector<Number> & initial_solution(*_nonlinear_implicit_system->solution);
+
+  auto & nodal_bcs = _nl.getPresetBCWarehouse();
+  const ConstBndNodeRange & bnd_nodes = _fe_problem.getCurrentAlgebraicBndNodeRange();
+  for (const auto & bnode : bnd_nodes)
+  {
+    BoundaryID boundary_id = bnode->_bnd_id;
+    Node * node = bnode->_node;
+    if (node->processor_id() == processor_id())
+    {
+      // reinit variables in nodes
+      _fe_problem.reinitNodeFace(node, boundary_id, 0);
+
+      if (nodal_bcs.hasActiveBoundaryObjects(boundary_id))
+      {
+        const auto & preset_bcs = nodal_bcs.getActiveBoundaryObjects(boundary_id);
+        for (const auto & preset_bc : preset_bcs)
+          preset_bc->computeValue(initial_solution);
+      }
+      if (nodal_bcs.hasActiveBoundaryObjects(boundary_id))
+      {
+        const auto & preset_bcs_res = nodal_bcs.getActiveBoundaryObjects(boundary_id);
+        for (const auto & preset_bc : preset_bcs_res)
+          preset_bc->computeValue(initial_solution);
+      }
+    }
+  }
+}
+
+void
+CentralDifferenceDirect::postResidual(NumericVector<Number> & residual)
+{
+  residual += _Re_time;
+  residual += _Re_non_time;
+  residual.close();
+
+  // Reset time to the time at which to evaluate nodal BCs, which comes next
+  _fe_problem.time() = _current_time;
+}
+
+bool
+CentralDifferenceDirect::performExplicitSolve(SparseMatrix<Number> & mass_matrix)
+{
+  bool converged = false;
+
+  // Computes the sum of each row (lumping)
+  // Note: This is actually how PETSc does it
+  // It's not "perfectly optimal" - but it will be fast (and universal)
+  mass_matrix.vector_mult(_mass_matrix_diag, *_ones);
+
+  // "Invert" the diagonal mass matrix
+  _mass_matrix_diag.reciprocal();
+
+  // Calculate acceleration
+  auto & accel = *_sys.solutionUDotDot();
+  accel.pointwise_mult(_mass_matrix_diag, _explicit_residual);
+
+  // Scaling the acceleration
+  auto accel_scaled = accel.clone();
+  accel_scaled->scale((_dt + _dt_old) / 2);
+
+  // Adding old vel to new vel
+  auto & vel = *_sys.solutionUDot();
+  const auto & old_vel = _sys.solutionUDotOld();
+  vel = *old_vel;
+  vel += *accel_scaled;
+
+  // Scale velocity by dt
+  auto vel_scaled = vel.clone();
+  vel_scaled->scale(_dt);
+  _solution_update = *vel_scaled;
+
+  vel.close();
+  accel.close();
+
+  // Check for convergence by seeing if there is a nan or inf
+  auto sum = _solution_update.sum();
+  converged = std::isfinite(sum);
+
+  // The linear iteration count remains zero
+  _n_linear_iterations = 0;
+  return converged;
```

Comment:
This is fine with me for now. In the long term, we need a better convergence check for explicit dynamics, or rather a _divergence check_.

Imagine we use what you have here as a convergence check: When the solve didn't not "converge", the framework handles solve failure, restores the solution, and travel back in time to reattempt the solve with a reduced time step. For implicit, that will work -- as we are moving closer to the basin of a traction. But for explicit, that doesn't help, in general.

Reply:
I agree, for now I was simply copying the current way of doing things found in `ExplicitTimeIntegrator.C`